### PR TITLE
Ban java serialization

### PR DESCRIPTION
**Pull request by s1monw**
We had several problems with Java Serialization in the past. At some point
in the Java 1.7.x series JDKs were not compatible anymore when java
serialization (ObjectStream) was used to exchange objects. In elasticsearch
we used this to serialize exceptions across the wire which caused several problems
with incompatible JDKs. While causing lot of trouble this essentially prevented
users from moving forward and upgrade their JVMs. To prevent these kind of issues
this commit removes the dependency on java serialization entirely and bans the
usage of ObjectOutputStream and ObjectInputStream entirely.

Yet, we can't fully serialize all exception anymore such that this commit
is best effort and adds hand written serialization to all elasticsearch exceptions
as well to a selected set of JDK and Lucene exceptions. (see StreamOutput#writeThrowable /
StreamInput.readThrowable). Stacktraces should be preserved for all exceptions while
several names might be replaced with ElasticsearchException if there is no mapping for
the given exception.

### CODE REVIEW

**Comment by rmuir** [comment]
Thanks for adding to forbidden apis! I like being more robust here, we don't want to discourage upgrading the JVM.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -290,4 +352,205 @@ public static String getExceptionName(Throwable ex) {
     public String toString() {
         return ExceptionsHelper.detailedMessage(this).trim();
     }
+
+    public static <T extends Throwable> T readStackTrace(T throwable, StreamInput in) throws IOException {
+        int stackTraceElements = in.readVInt();
+        StackTraceElement[] stackTrace = new StackTraceElement[stackTraceElements];
+        for (int i = 0; i < stackTraceElements; i++) {
+            String declaringClasss = in.readString();
+            String fielName = in.readString();
```
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: small typo

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 2]
```diff
@@ -445,8 +451,46 @@ public void writeOptionalStreamable(@Nullable Streamable streamable) throws IOEx
...
+                writeVInt(6);
+            } else if (throwable instanceof IllegalStateException) {
+                writeVInt(7);
+            } else if (throwable instanceof EOFException) {
+                writeVInt(8);
+            } else if (throwable instanceof SecurityException) {
+                writeVInt(9);
+            } else {
+                ElasticsearchException ex;
+                if (throwable instanceof ElasticsearchException) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: How is this case possible given the first if clause above?

**Comment by rjernst** [comment]
LGTM, one small question.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: yeah that was a bug - fixed it in the last push ;)... I added unittests FTW

**Comment by s1monw** [comment]
I added a bunch of unittests for all the custom serialization exceptions I think we are pretty close here... I would like to move forward with this and push this tomorrow unless anybody objects.

**Comment by kimchy** [comment]
@s1monw small comment since I pulled this change and noticed, the check in StreamOutput#writeThrowable for `AlreadyClosedException` is not going to execute since it extends IllegalStateException which is checked before.

**Comment by s1monw** [comment]
> @s1monw small comment since I pulled this change and noticed, the check in StreamOutput#writeThrowable for AlreadyClosedException is not going to execute since it extends IllegalStateException which is checked before.

pushed a fix - I had a test for it but it was jagged!

**Comment by vvcephei** [comment]
hey @s1monw, I guess you missed the discussion of this: https://github.com/vvcephei/es-rest-client-java in https://github.com/elastic/elasticsearch/issues/7743, and I missed this PR!

Someone just told me about this project, so I haven't looked at it yet, but I'm super stoked you're doing this. If it helps you, my project has delete, get, index, and search fully implemented. We are working on deserializing exceptions now.

Are you planning to support 1.x versions at all? If not, I may just re-label my project as a 1.x client. If you are going to support 1.x apis, I'd be interested in seeing if I can help you out at all.

**Pull request merged**